optout/s3_file_handler.go

```diff
@@ -1 +1,220 @@
+	// Optional S3 endpoint to use for connection.
+	Endpoint string
+	// Optional role to assume when connecting to S3.
+	AssumeRoleArn string
+}
+
+// Define logger functions to ensure that logs get sent to:
+// 1. Splunk (Logger.*)
+// 2. stdout (Jenkins)
+
+func (handler *S3FileHandler) Infof(format string, rest ...interface{}) {
+	fmt.Printf(format, rest...)
+	handler.Logger.Infof(format, rest...)
+}
+
+func (handler *S3FileHandler) Warningf(format string, rest ...interface{}) {
+	fmt.Printf(format, rest...)
+	handler.Logger.Warningf(format, rest...)
+}
+
+func (handler *S3FileHandler) Errorf(format string, rest ...interface{}) {
+	fmt.Printf(format, rest...)
+	handler.Logger.Errorf(format, rest...)
+}
+
+func (handler *S3FileHandler) LoadOptOutFiles(path string) (suppressList *[]*OptOutFilenameMetadata, skipped int, err error) {
+	var result []*OptOutFilenameMetadata
+
+	bucket, prefix, err := parseS3Uri(path)
+	if err != nil {
+		handler.Errorf("Failed to parse S3 path: %s", err)
+		return &result, skipped, err
+	}
+
+	sess, err := handler.createSession()
+	if err != nil {
+		handler.Errorf("Failed to create S3 session: %s", err)
+		return &result, skipped, err
+	}
+
+	svc := s3.New(sess)
+
+	handler.Infof("Listing objects in bucket %s, prefix %s", bucket, prefix)
+
+	resp, err := svc.ListObjects(&s3.ListObjectsInput{
+		Bucket: aws.String(bucket),
+		Prefix: aws.String(prefix),
+	})
+
+	if err != nil {
+		handler.Errorf("Failed to list objects in S3 bucket %s, prefix %s: %s", bucket, prefix, err)
+		return &result, skipped, err
+	}
+
+	for _, obj := range resp.Contents {
+		metadata, err := ParseMetadata(*obj.Key)
+		metadata.FilePath = fmt.Sprintf("s3://%s/%s", bucket, *obj.Key)
+		metadata.DeliveryDate = *obj.LastModified
+
+		if err != nil {
+			// Skip files with a bad name.  An unknown file in this dir isn't a blocker
+			handler.Warningf("Unknown file found: %s. Skipping.", metadata)
+			skipped = skipped + 1
+			continue
+		}
+
+		result = append(result, &metadata)
+	}
+
+	return &result, skipped, err
+}
+
+func (handler *S3FileHandler) OpenFile(metadata *OptOutFilenameMetadata) (*bufio.Scanner, func(), error) {
+	handler.Infof("Opening file %s", metadata.FilePath)
+	bucket, file, err := parseS3Uri(metadata.FilePath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sess, err := handler.createSession()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	downloader := s3manager.NewDownloader(sess)
+	buff := &aws.WriteAtBuffer{}
+	numBytes, err := downloader.Download(buff, &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(file),
+	})
+
+	if err != nil {
+		handler.Errorf("Failed to download bucket %s, key %s", bucket, file)
+		return nil, nil, err
+	}
+
+	handler.Infof("file downloaded: size=%d", numBytes)
+
+	byte_arr := buff.Bytes()
+	sc := bufio.NewScanner(bytes.NewReader(byte_arr))
+	return sc, func() {}, err
+}
+
+func (handler *S3FileHandler) CleanupOptOutFiles(suppresslist []*OptOutFilenameMetadata) error {
```

Comment:
Are these opt-out files per-consumer, or shared between them?

Reply:
Per consumer (each API team has their own S3 folders that only they have access to.)